src/D2L.CodeStyle.Analyzers/Immutability/AddImmutableExceptionsCodeFix.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[ExportCodeFixProvider(
+		LanguageNames.CSharp,
+		Name = nameof( AddImmutableExceptionsCodeFix )
+	)]
+	public sealed class AddImmutableExceptionsCodeFix : CodeFixProvider {
+		public override ImmutableArray<string> FixableDiagnosticIds
+			=> ImmutableArray.Create(
+				Diagnostics.ImmutableExceptionInheritanceIsInvalid.Id
+			);
+
+		public override FixAllProvider GetFixAllProvider() {
+			return WellKnownFixAllProviders.BatchFixer;
+		}
+
+		public override async Task RegisterCodeFixesAsync(
+			CodeFixContext context
+		) {
+			var root = await context.Document
+				.GetSyntaxRootAsync( context.CancellationToken )
+				.ConfigureAwait( false );
+
+			var model = await context.Document
+				.GetSemanticModelAsync( context.CancellationToken )
+				.ConfigureAwait( false );
+
+			foreach( var diagnostic in context.Diagnostics ) {
+				var attr = root.FindNode(
+					diagnostic.Location.SourceSpan
+				) as AttributeSyntax;

```

Comment:
True. Though it kind of depends on the asynchronicity of this. Since the location is essentially just a pointer into the file, the file changing while the codefix is running could cause the location to be out-of-date and no longer point to the attribute. I _think_... I could be mistaken on the mechanics involved here. Just seems like a better safe than sorry kind of thing. :)

Reply:
Ok no NRE happens because `GetImmutableAttributeSyntax` always returns a non-null things. It has a trap at the bottom:

https://github.com/Brightspace/D2L.CodeStyle/blob/626d26ffe915e29a64cdbc0fb24a57e70826a1ac/src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityExceptionInheritanceAnalyzer.cs#L151-L152

I was going to explain why this never gets hit but I'm wrong: we're taking a particular decl syntax, which could be a partial class, and looking for the attribute _syntax_ attached to the declaration syntax. The `[Immutable]` attribute syntax might actually be on a different piece of syntax (i.e. another partial class syntax.) So we wouldn't find it. This needs some rework. Thanks.

Reply:
Oh I missed your last comment. I was kind of assuming that we got a snapshot of the syntax tree. Maybe that's not true...

Reply:
Slack converation:

>**jacob.parker** [2:42 PM]
> hey, so the immutability exception analyzer:
> 
> right now it looks at every decl (class interface or struct)
> this will cause it to emit one diagnostic for every decl syntax. so a partial class with multiple decls could get multiple diagnostics for the same problem.
> 
> i'd like to change it to instead analyze `AttributeSyntax` nodes and look for `[Immutable]`. the other PR i have open puts the diagnostic at the `AttributeSyntax` anyway, and that will cause an exception as you pointed out because e.g. for partial classes there will only be an attribute attached to one decl syntax.
> 
> this wouldn't have worked before because we've needed to analyze types that don't have `[Immutable]` explicitly (which today means they get all the exceptions) so might not have an `AttributeSyntax` that'll get picked up. however now we are requiring that `[Immutable]` is always explicit so it should be good.
> 
> does that make sense?
> 
> **matt.jones** [2:45 PM]
> As long as the new explicit `[Immutable]` requirement sticks around, I think that makes sense, yup.

since the problem here (an exception if the code fix was applied to a partial class) is rare enough to not impact us currently I'm going to merge this and work on the fix on top of these changes.